File: packages/dotnet/src/Yourdudeken.Mpesa/Services/STKPushService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Yourdudeken.Mpesa.Config;
using Yourdudeken.Mpesa.Http;
using Yourdudeken.Mpesa.Auth;

namespace Yourdudeken.Mpesa.Services
{
    public class STKPushService
    {
        private readonly HttpClientWrapper _httpClient;
        private readonly AuthService _auth;
        private readonly MpesaConfig _config;
        private readonly string _baseUrl;

        public STKPushService(HttpClientWrapper httpClient, AuthService auth, MpesaConfig config)
        {
            _httpClient = httpClient;
            _auth = auth;
            _config = config;
            _baseUrl = httpClient.GetBaseUrl();
        }

        public async Task<string> Push(string phonenumber, int amount, string accountNumber,
            string callbackUrl = null, string transactionType = "CustomerPayBillOnline", string shortCodeType = "C2B")
        {
            if (string.IsNullOrEmpty(accountNumber))
                throw new ArgumentException("An Account Reference is required for All transactions.");

            var url = $"{_baseUrl}/mpesa/stkpush/v1/processrequest";
            var data = new Dictionary<string, object>
            {
                { "BusinessShortCode", _config.Shortcode },
                { "Password", LipaNaMpesaPassword() },
                { "Timestamp", GetFormattedTimestamp() },
                { "Amount", amount },
                { "PartyA", PhoneValidator(phonenumber) },
                { "PartyB", transactionType == "CustomerPayBillOnline" ? _config.Shortcode : _config.TillNumber },
                { "TransactionType", transactionType },
                { "PhoneNumber", PhoneValidator(phonenumber) },
                { "TransactionDesc", "Payment" },
                { "AccountReference", accountNumber },
                { "CallBackURL", callbackUrl ?? _config.CallbackUrl }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, data, token);
        }

        public async Task<string> Query(string checkoutRequestId, string shortCodeType = "C2B")
        {
            var url = $"{_baseUrl}/mpesa/stkpushquery/v1/query";
            var data = new Dictionary<string, object>
            {
                { "BusinessShortCode", _config.Shortcode },
                { "Password", LipaNaMpesaPassword() },
                { "Timestamp", GetFormattedTimestamp() },
                { "CheckoutRequestID", checkoutRequestId }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, data, token);
        }

        private string PhoneValidator(string phoneNo)
        {
            if (phoneNo.StartsWith("+")) phoneNo = phoneNo.Substring(1);
            if (phoneNo.StartsWith("0")) phoneNo = "254" + phoneNo.Substring(1);
            else if (phoneNo.StartsWith("7")) phoneNo = "254" + phoneNo;
            return phoneNo;
        }

        private string GetFormattedTimestamp() => DateTime.Now.ToString("yyyyMMddHHmmss");

        private string LipaNaMpesaPassword()
        {
            var timestamp = GetFormattedTimestamp();
            var password = _config.Shortcode + _config.Passkey + timestamp;
            return Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(password));
        }
    }
}

File: packages/java/src/main/java/com/yourdudeken/mpesa/services/STKPushService.java
package com.yourdudeken.mpesa.services;

import com.yourdudeken.mpesa.config.MpesaConfig;
import com.yourdudeken.mpesa.http.HttpClient;
import com.yourdudeken.mpesa.auth.AuthService;
import java.util.HashMap;
import java.util.Map;

public class STKPushService {
    private HttpClient httpClient;
    private AuthService auth;
    private MpesaConfig config;
    private String baseUrl;

    public STKPushService(HttpClient httpClient, AuthService auth, MpesaConfig config) {
        this.httpClient = httpClient;
        this.auth = auth;
        this.config = config;
        this.baseUrl = httpClient.getBaseUrl();
    }

    public String push(String phonenumber, int amount, String accountNumber, 
                      String callbackUrl, String transactionType, String shortCodeType) {
        if (accountNumber == null || accountNumber.isEmpty()) {
            throw new IllegalArgumentException("An Account Reference is required for All transactions.");
        }

        String url = baseUrl + "/mpesa/stkpush/v1/processrequest";
        Map<String, Object> data = new HashMap<>();
        data.put("BusinessShortCode", config.getShortcode());
        data.put("Password", lipaNaMpesaPassword());
        data.put("Timestamp", getFormattedTimestamp());
        data.put("Amount", amount);
        data.put("PartyA", phoneValidator(phonenumber));
        data.put("PartyB", transactionType.equals("CustomerPayBillOnline") 
            ? config.getShortcode() : config.getTillNumber());
        data.put("TransactionType", transactionType);
        data.put("PhoneNumber", phoneValidator(phonenumber));
        data.put("TransactionDesc", "Payment");
        data.put("AccountReference", accountNumber);
        data.put("CallBackURL", callbackUrl != null ? callbackUrl : config.getCallbackUrl());

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, data, token);
    }

    public String query(String checkoutRequestId, String shortCodeType) {
        String url = baseUrl + "/mpesa/stkpushquery/v1/query";
        Map<String, Object> data = new HashMap<>();
        data.put("BusinessShortCode", config.getShortcode());
        data.put("Password", lipaNaMpesaPassword());
        data.put("Timestamp", getFormattedTimestamp());
        data.put("CheckoutRequestID", checkoutRequestId);

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, data, token);
    }

    private String phoneValidator(String phoneNo) {
        if (phoneNo.startsWith("+")) phoneNo = phoneNo.substring(1);
        if (phoneNo.startsWith("0")) phoneNo = "254" + phoneNo.substring(1);
        else if (phoneNo.startsWith("7")) phoneNo = "254" + phoneNo;
        return phoneNo;
    }

    private String getFormattedTimestamp() {
        java.time.LocalDateTime now = java.time.LocalDateTime.now();
        return String.format("%04d%02d%02d%02d%02d%02d",
            now.getYear(), now.getMonthValue(), now.getDayOfMonth(),
            now.getHour(), now.getMinute(), now.getSecond());
    }

    private String lipaNaMpesaPassword() {
        String timestamp = getFormattedTimestamp();
        String password = config.getShortcode() + config.getPasskey() + timestamp;
        return java.util.Base64.getEncoder().encodeToString(password.getBytes());
    }
}
